security/pkg/pki/custom/ca.go

```diff
@@ -0,0 +1,174 @@
+// Copyright Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package custom
+
+import (
+	"context"
+	"crypto/tls"
+	"crypto/x509"
+	"encoding/pem"
+	"fmt"
+	"io/ioutil"
+	"time"
+
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials"
+
+	pb "istio.io/api/security/v1alpha1"
+	"istio.io/istio/security/pkg/pki/util"
+	"istio.io/pkg/log"
+)
+
+var cLog = log.RegisterScope("CustomCAClient", "Custom CA Integration Log", 0)
+
+const (
+	connectTimeout = 30 * time.Second
+)
+
+// CAClientOpts options
+type CAClientOpts struct {
+	CAAddr         string
+	RootCertPath   string
+	ClientKeyPath  string
+	ClientCertPath string
+	KeyCertBundle  util.KeyCertBundle
+	RequestTimeout int
+}
+
+// CAClient generates keys and certificates for Istio identities.
+type CAClient struct {
+	opts         *CAClientOpts
+	caClientConn *grpc.ClientConn
+	pbClient     pb.IstioCertificateServiceClient
+}
+
+// NewCAClient returns a new CAClient instance.
+func NewCAClient(opts *CAClientOpts) (*CAClient, error) {
```

Comment:
maybe could merge with citadel client

Reply:
Thank @hzxuzhonghu ,
I'll have a future work to merge this with Citadel client. Currently, It should be separated to minimize effect on things that are stabilizing.

Reply:
sgtm

Reply:
Maybe a TODO ?